Add more REST API docs (Dashboard Widget Data, Dynamic Content, Point Actions, Point Triggers, Webhooks) (#144)
mautic/developer-documentation-new: docs/rest_api/dynamic_content.rst

Dynamic Content
###############

> Use this endpoint to obtain details on Mautic's web Dynamic Content.

Reviewer: ```suggestion
Use this endpoint to obtain details on Mautic's Dynamic Web Content.
```

     - Only return currently published entities.
   * - ``minimal``
     - Return only array of entities without additional lists in it.

**Response**

``Expected Response Code: 200``

.. code-block:: json

   {
       "total":30,
       "dynamicContents":[
           {
               "isPublished":true,
               "dateAdded":"2016-06-20T11:27:09+00:00",
               "createdBy":1,
               "createdByUser":"John Doe",
               "dateModified":"2016-08-22T17:14:01+00:00",
               "modifiedBy":1,
               "modifiedByUser":"John Doe",
               "id":2,
               "name":"CD2",
               "category":null,
               "publishUp":null,
               "publishDown":null,
               "sentCount":0,
               "variantParent":null,
               "variantChildren":[]
           }
       ]
   }

**Properties**

Same as `Get Dynamic Content <#get-dynamic-content>`_.

.. vale off

Create Dynamic Content
**********************

.. vale on

.. code-block:: php

   <?php

   $data = array(
       'name'        => 'Dynamic Content A',
       'isPublished' => 1
   );

   $dynamicContent = $dynamicContentApi->create($data);

> Create a new dynamicContent.

Reviewer: ```suggestion
Create a new Dynamic Content.
```
I _think_ these should be separate words unless it's purposely written as above in the code sample.

     - Name of the User that created the Dynamic Content
   * - ``dateModified``
     - datetime/null
     - Date/time Dynamic Content was last modified
   * - ``modifiedBy``
     - int
     - ID of the User that last modified the Dynamic Content
   * - ``modifiedByUser``
     - string
     - Name of the User that last modified the Dynamic Content
   * - ``variantChildren``
     - array
     - Array of Dynamic Content entities for variants of this landing Dynamic Content
   * - ``variantParent``
     - object
     - The parent/main Dynamic Content if this is a variant, also known as A/B test
   * - ``sentCount``
     - int
     - Count of how many times the Dynamic Content got sent

**Response**

``Expected Response Code: 201``

**Properties**

Same as `Get Dynamic Content <#get-dynamic-content>`_.

.. vale off

Edit Dynamic Content
********************

.. vale on

.. code-block:: php

   <?php

   $id   = 1;
   $data = array(
       'name'        => 'New dynamicContent name',
       'isPublished' => 0
   );

   // Create new a dynamicContent of ID 1 isn't found?
   $createIfNotFound = true;

   $dynamicContent = $dynamicContentApi->edit($id, $data, $createIfNotFound);

> Edit a new dynamicContent. Note that this supports PUT or PATCH depending on the desired behavior.

Reviewer: ```suggestion
Edit a new Dynamic Content. Note that this supports PUT or PATCH depending on the desired behavior.
```
As above.

> **PUT** creates a dynamicContent if the given ID doesn't exist and clears all the Dynamic Content information, adds the information from the request.

Reviewer: ```suggestion
**PUT** creates a Dynamic Content if the given ID doesn't exist and clears all the Dynamic Content information, adds the information from the request.
```

**PATCH** fails if the Dynamic Content with the given ID doesn't exist and updates the Dynamic Content field values with the values from the request.

.. vale off

**HTTP Request**

.. vale on

> To edit a dynamicContent and return a 404 if the Dynamic Content isn't found:

Reviewer: ```suggestion
To edit a Dynamic Content and return a 404 if the Dynamic Content isn't found:
```

``PATCH /dynamiccontents/ID/edit``

> To edit a dynamicContent and create a new one if the Dynamic Content isn't found:

Reviewer: ```suggestion
To edit a Dynamic Content and create a new one if the Dynamic Content isn't found:
```

   * - ``createdByUser``
     - string
     - Name of the User that created the Dynamic Content
   * - ``dateModified``
     - datetime/null
     - Date/time Dynamic Content was last modified
   * - ``modifiedBy``
     - int
     - ID of the User that last modified the Dynamic Content
   * - ``modifiedByUser``
     - string
     - Name of the User that last modified the Dynamic Content
   * - ``variantChildren``
     - array
     - Array of Dynamic Content entities for variants of this landing Dynamic Content
   * - ``variantParent``
     - object
     - The parent/main Dynamic Content if this is a variant, also known as A/B test
   * - ``sentCount``
     - int
     - Count of how many times the Dynamic Content got sent

**Response**

If ``PUT``, the expected response code is ``200`` if editing an existing Dynamic Content entry or ``201`` if creating a new one.

If ``PATCH``, the expected response code is ``200``.

**Properties**

Same as `Get Dynamic Content <#get-dynamic-content>`_.

.. vale off

Delete Dynamic Content
**********************

.. vale on

.. code-block:: php

   <?php

   $dynamicContent = $dynamicContentApi->delete($id);

> Delete a dynamicContent.

Reviewer: ```suggestion
Delete a Dynamic Content.
```
